mdata/notifierKafka/notifierKafka.go

```diff
@@ -176,36 +173,57 @@ func (c *NotifierKafka) flush() {
 		return
 	}
 
-	msg := mdata.PersistMessageBatch{Instance: c.instance, SavedChunks: c.buf}
-	c.buf = nil
-
-	go func() {
-		log.Debug("kafka-notifier sending %d batch metricPersist messages", len(msg.SavedChunks))
-
-		data, err := json.Marshal(&msg)
+	// In order to correctly routes the saveMessages to the correct partition,
+	// we cant send them in batches anymore.
+	payload := make([]*sarama.ProducerMessage, 0, len(c.buf))
+	var pMsg mdata.PersistMessageBatch
+	for i, msg := range c.buf {
+		def, err := c.idx.Get(strings.SplitN(msg.Key, "_", 2)[0])
 		if err != nil {
-			log.Fatal(4, "kafka-notifier failed to marshal persistMessage to json.")
+			log.Error(3, "kafka-cluster: failed to lookup metricDef with id %s", msg.Key)
+			continue
 		}
-		buf := new(bytes.Buffer)
+		buf := bytes.NewBuffer(c.bPool.Get())
 		binary.Write(buf, binary.LittleEndian, uint8(mdata.PersistMessageBatchV1))
-		buf.Write(data)
+		encoder := json.NewEncoder(buf)
+		pMsg = mdata.PersistMessageBatch{Instance: c.instance, SavedChunks: c.buf[i:i]}
```

Comment:
i'm confused about `[i:i]`, wouldn't that always return an empty list? should that be `[i:i+1]`?